### PR TITLE
Store Quest Progress Per Player in Individual Files

### PR DESCRIPTION
**Pull request by TechLord22**
GTNewHorizons#61

### CODE REVIEW

File: src/main/java/betterquesting/api2/storage/INBTProgress.java [hunk 1]
```diff
@@ -8,7 +8,14 @@
 
 // Used when progress specific data is being handled (usually split per user)
 public interface INBTProgress<T extends NBTBase> {
+
```
Comment (Exaxxion) [inline src/main/java/betterquesting/api2/storage/INBTProgress.java:13]: I would recommend revising these comment blocks to be proper Javadocs. The descriptions provided are essentially just `@param` of one parameter of each method rather than a complete documentation of the function.

File: src/main/java/betterquesting/api/events/MarkDirtyPlayerEvent.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.Collection;
+import java.util.Collections;
+import java.util.TreeSet;
+import java.util.UUID;
+
+public class MarkDirtyPlayerEvent extends Event {
+
+    private final Collection<UUID> dirtyPlayerIDs;
+
+    public MarkDirtyPlayerEvent(UUID dirtyPlayerID) {
```
Comment (Exaxxion) [inline src/main/java/betterquesting/api/events/MarkDirtyPlayerEvent.java:14]: I believe we may need to be cautious about nulls here. `Collections` methods like `singleton(T o)` will throw NPE for a null argument, and `TreeSet` will similarly reject adding null values.

File: src/main/java/betterquesting/handlers/SaveLoadHandler.java [hunk 3]
```diff
@@ -74,67 +89,126 @@ public void loadDatabases(MinecraftServer server) {
...
+
+        allFutures.add(saveParties());
+
+        allFutures.add(saveNames());
+
+        allFutures.add(saveLives());
+
+        MinecraftForge.EVENT_BUS.post(new DatabaseEvent.Save(DBType.ALL));
+
+        for (Future<Void> future : allFutures) {
```
Comment (Exaxxion) [inline src/main/java/betterquesting/handlers/SaveLoadHandler.java:138]: If the saving was interrupted or failed, do we really want to clear the `isDirty` flag? Aren't there still unsaved changes?

File: src/main/java/betterquesting/handlers/SaveLoadHandler.java [hunk 4]
```diff
@@ -74,67 +89,126 @@ public void loadDatabases(MinecraftServer server) {
...
+            NameCache.INSTANCE.reset();
+            PartyInvitations.INSTANCE.reset();
+            PartyManager.INSTANCE.reset();
+
+            if (BetterQuesting.proxy.isClient()) {
+                GuiHome.bookmark = null;
+                QuestNotification.resetNotices();
+            }
+
+            // TODO: Fire an event to that expansions can use to reset their own databases if necessary
```
Comment (Exaxxion) [inline src/main/java/betterquesting/handlers/SaveLoadHandler.java:169]: If you're adding TODO comments then I would argue that this PR is incomplete. Get rid of the comment and open an issue ticket to track the feature, or implement the behavior you're reminding yourself to implement.

File: src/main/java/betterquesting/handlers/SaveLoadHandler.java [hunk 5]
```diff
@@ -74,67 +89,126 @@ public void loadDatabases(MinecraftServer server) {
...
+        });
+    }
+
+    private void loadConfig() {
+        QuestSettings.INSTANCE.reset();
+        QuestDatabase.INSTANCE.reset();
+        QuestLineDatabase.INSTANCE.reset();
 
         boolean useDef = !fileDatabase.exists();
```
Comment (Exaxxion) [inline src/main/java/betterquesting/handlers/SaveLoadHandler.java]: The logic around this `useDef` parameter seems unnecessarily convoluted.

Why not just do like:
```java

if(!fileDatabase.exists()) {
    isDirty = true;
    fileDatabase = fileDefaultDatabase;
} else {
// [...] skipping to line 197
JsonObject json = JsonHelper.ReadFromFile(fileDatabase);
```

File: src/main/java/betterquesting/handlers/SaveLoadHandler.java [hunk 6]
```diff
@@ -74,67 +89,126 @@ public void loadDatabases(MinecraftServer server) {
...
-        if (!cVer.equalsIgnoreCase(bVer) && !useDef) // RUN BACKUPS
+        if (!currVer.equalsIgnoreCase(buildVer) && !useDef) // RUN BACKUPS
         {
-            String fsVer = JsonHelper.makeFileNameSafe(bVer);
+            String fsVer = JsonHelper.makeFileNameSafe(buildVer);
 
             if (fsVer.length() <= 0) fsVer = "pre-251";
 
-            BetterQuesting.logger.warn("BetterQuesting has been updated to from \"" + fsVer + "\" to \"" + cVer + "\"! Creating back ups...");
+            BetterQuesting.logger.warn("BetterQuesting has been updated to from \"" + fsVer + "\" to \"" + currVer + "\"! Creating backups...");
```
Comment (Exaxxion) [inline src/main/java/betterquesting/handlers/SaveLoadHandler.java]: We use log4j; use parameter substitution instead of string concatenation.

File: src/main/java/betterquesting/handlers/SaveLoadHandler.java [hunk 7]
```diff
@@ -143,143 +217,139 @@ public void loadDatabases(MinecraftServer server) {
...
-        // === NAMES ===
-
-        JsonObject j4 = JsonHelper.ReadFromFile(fileNames);
-
-        NBTTagCompound nbt4 = NBTConverter.JSONtoNBT_Object(j4, new NBTTagCompound(), true);
-        NameCache.INSTANCE.readFromNBT(nbt4.getTagList("nameCache", 10), false);
-
-        // === LIVES ===
+        getPlayerProgressFiles().forEach(file -> {
+            JsonObject json = JsonHelper.ReadFromFile(file);
```
Comment (Exaxxion) [inline src/main/java/betterquesting/handlers/SaveLoadHandler.java:243]: If there's any chance of an exception being thrown in this block it could result in a really ugly stack trace. I would honestly suggest just doing a standard for-each loop here.

File: src/main/java/betterquesting/handlers/SaveLoadHandler.java [hunk 8]
```diff
@@ -143,143 +217,139 @@ public void loadDatabases(MinecraftServer server) {
...
-        NBTTagCompound jsonN = new NBTTagCompound();
+        return JsonHelper.WriteToFile(fileDatabase, NBTConverter.NBTtoJSON_Compound(json, new JsonObject(), true));
+    }
 
-        jsonN.setTag("nameCache", NameCache.INSTANCE.writeToNBT(new NBTTagList(), null));
+    private List<Future<Void>> saveProgress() {
+        final List<Future<Void>> futures = dirtyPlayers.stream().map(this::savePlayerProgress).collect(Collectors.toList());
+        dirtyPlayers.clear();
+        return futures;
+    }private Future<Void> saveParties() {
```
Comment (Exaxxion) [inline src/main/java/betterquesting/handlers/SaveLoadHandler.java]: looks like a missing newline here?

File: src/main/java/betterquesting/handlers/SaveLoadHandler.java [hunk 9]
```diff
@@ -143,143 +217,139 @@ public void loadDatabases(MinecraftServer server) {
...
+        LifeDatabase.INSTANCE.readFromNBT(nbt.getCompoundTag("lifeDatabase"), false);
+    }
 
-            jsonCon.setTag("questSettings", QuestSettings.INSTANCE.writeToNBT(new NBTTagCompound()));
-            jsonCon.setTag("questDatabase", QuestDatabase.INSTANCE.writeToNBT(new NBTTagList(), null));
-            jsonCon.setTag("questLines", QuestLineDatabase.INSTANCE.writeToNBT(new NBTTagList(), null));
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    private void checkLegacyFiles(File rootDir) {
+        if (new File(rootDir, "QuestDatabase.json").exists() && !fileDatabase.exists()) {
+            File legFileDat = new File(rootDir, "QuestDatabase.json");
```
Comment (Exaxxion) [inline src/main/java/betterquesting/handlers/SaveLoadHandler.java]: Nomenclature here is awkward. "File" is redundant since that's the type of the variable. Three letter abbreviations are hard to follow. Why not just "legacyQuests", "legacyProgress", etc.?

File: src/main/java/betterquesting/handlers/SaveLoadHandler.java (hunk 4 above)
Comment (ALongStringOfNumbers) [reply src/main/java/betterquesting/handlers/SaveLoadHandler.java:169]: I believe this TODO existed initially in the file, but was just moved around.

Comment (Exaxxion) [reply src/main/java/betterquesting/handlers/SaveLoadHandler.java:169]: That could well be, I think I recall seeing it much later in the file. This was one of the first comments I drafted and I spent about an hour reviewing the code.

**Comment by Exaxxion** [comment]
I'd say that my review concerns have all been suitably addressed. Only thing further I have to add is perhaps consider `Optional` for null-coalescing in the `MarkDirtyPlayerEvent` constructors.
